src/lib/Timer.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * General purpose time recording tool
+ */
+class Timer {
+    /** @property {Object} - Dictionary of time objects. Used to track times. tag: { start, end, elapsed } */
+    times = {};
+
+    /**
+     * Start running a timer, for a tag. If a tag has already been started, then this does nothing.
+     *
+     * @public
+     * @param {string} tag - Time structure to look up, at this string. Creates a new one if none exists.
+     * @return {void}
+     */
+    start(tag) {
+        let time = this.get(tag);
+        if (!time) {
+            time = this.create(tag);
+        }
+
+        // Can't start a timer that's already started
+        if (time.start) {
+            return;
+        }
+
+        time.start = global.performance.now();
+        time.end = undefined;
+        time.elapsed = undefined;
+    }
+
+    /**
+     * Stop the timer from running, for a tag, and calculate elapsed time. If a tag has never been started, or has already
+     * been stopped, this does nothing.
+     *
+     * @public
+     * @param {string} tag - Time structure to look up, at this string.
+     * @return {void}
+     */
+    stop(tag) {
+        const time = this.get(tag);
+        // The timer has already been stopped, or hasn't started. Don't stop it again.
+        if (!time || time.start === undefined || time.end !== undefined) {
+            return;
+        }
+
+        time.end = global.performance.now();
+        time.elapsed = time.end - time.start;
+    }
+
+    /**
+     * Get the time object at a given tag.
+     *
+     * @public
+     * @param {string} tag - The tag to get the time structure for.
+     * @return {Object} The time structure, or undefined if none with that tag.
+     */
+    get(tag) {
+        return this.times[tag];
+    }
+
+    /**
+     * Resets the values in a certain time structure, if it exists.
+     *
+     * @public
+     * @param {string} [tag] - If provided, will reset a specific time structure associated with the tag.
+     *                         If empty, resets everything.
+     * @return {void}
+     */
+    reset(tag) {
+        if (tag) {
+            const time = this.get(tag);
+            // We don't need to clean up nothin'
+            if (!time) {
+                return;
+            }
+
+            time.start = undefined;
+            time.end = undefined;
+            time.elapsed = undefined;
+        } else {
+            Object.keys(this.times).forEach((timeTag) => {
```

Comment:
Object.keys(this.times).forEach(this.reset);

Reply:
boom, done